File: carrera-sdk/producer/go/src/carrera/messageBuilder.go
package carrera

import (
	"carrera/common"
	"carrera/CarreraProducer"
	"carrera/common/util"
)

type MessageBuilder struct {
	producer sender
	msg      *CarreraProducer.Message
}

func NewMessageBuilder(producer sender) *MessageBuilder {
	this := &MessageBuilder{}
	this.initMsg(producer)

	return this
}

func (this *MessageBuilder) initMsg(producer sender) {
	this.msg = CarreraProducer.NewMessage()
	this.producer = producer
	this.msg.Properties = make(map[string]string)
	curSdkVersion := common.CUR_SDK_VERSION
	this.msg.Version = &curSdkVersion
	this.msg.PartitionId = PARTITION_RAND
}

func (this *MessageBuilder) SetTopic(topic string) *MessageBuilder {
	this.msg.Topic = topic
	return this
}

func (this *MessageBuilder) SetPartitionId(partitionId int32) *MessageBuilder {
	this.msg.PartitionId = partitionId
	return this
}

func (this *MessageBuilder) SetRandomPartition() *MessageBuilder {
	this.msg.PartitionId = PARTITION_RAND
	return this
}

func (this *MessageBuilder) SetHashId(hashId int64) *MessageBuilder {
	this.msg.PartitionId = PARTITION_HASH
	this.msg.HashId = hashId
	return this
}

func (this *MessageBuilder) SetBody(body interface{}) *MessageBuilder {
	if body != nil {
		switch b := body.(type) {
		case []byte:
			this.msg.Body = b
		case string:
			this.msg.Body = []byte(b)
		}
	}
	return this
}

func (this *MessageBuilder) SetKey(key string) *MessageBuilder {
	if key != "" {
		this.msg.Key = key
	}
	return this
}

func (this *MessageBuilder) SetTags(tag string) *MessageBuilder {
	if tag != "" {
		this.msg.Tags = tag
	}

	return this
}

func (this *MessageBuilder) SetTraceId(traceId string) *MessageBuilder {
	this.msg.Properties[CarreraProducer.TRACE_ID] = traceId
	return this
}

func (this *MessageBuilder) SetSpanId(spanId string) *MessageBuilder {
	this.msg.Properties[CarreraProducer.SPAN_ID] = spanId
	return this
}

func (this *MessageBuilder) SetPressureTraffic(isOpen bool) *MessageBuilder {
	if isOpen {
		this.msg.Properties[CarreraProducer.PRESSURE_TRAFFIC_KEY] = CarreraProducer.PRESSURE_TRAFFIC_ENABLE
	} else {
		this.msg.Properties[CarreraProducer.PRESSURE_TRAFFIC_KEY] = CarreraProducer.PRESSURE_TRAFFIC_DISABLE
	}
	return this
}

func (this *MessageBuilder) AddProperty(key, value string) *MessageBuilder {
	if key != "" && key != CarreraProducer.PRESSURE_TRAFFIC_KEY {
		this.msg.Properties[key] = value
	}
	return this
}

func (this *MessageBuilder) Send() *CarreraProducer.Result {
	if this.msg.Key == "" {
		this.SetKey(util.GenRandKey())
	}
	return this.producer.sendMessage(this.msg)
}

File: carrera-sdk/producer/java/carrera-producer-sdk/src/main/java/com/xiaojukeji/carrera/producer/MessageBuilder.java
package com.xiaojukeji.carrera.producer;

import com.xiaojukeji.carrera.config.CarreraConfig;
import com.xiaojukeji.carrera.thrift.Message;
import com.xiaojukeji.carrera.thrift.Result;
import com.xiaojukeji.carrera.utils.VersionUtils;
import com.xiaojukeji.carrera.thrift.producerProxyConstants;

import java.io.UnsupportedEncodingException;
import java.util.HashMap;

import com.xiaojukeji.carrera.utils.ConstUtils;
import com.xiaojukeji.carrera.utils.RandomKeyUtils;


public class MessageBuilder {
    private CarreraProducer producer = null;
    private Message message = null;

    public MessageBuilder(CarreraProducer producer) {
        this.producer = producer;
        this.message = new Message();
        this.message.setVersion(VersionUtils.getVersion());
        this.message.setProperties(new HashMap<String, String>());
        this.message.partitionId = ConstUtils.PARTITION_RAND;
    }

    public MessageBuilder setTopic(String topic) {
        this.message.setTopic(topic);
        return this;
    }

    public MessageBuilder setPartitionId(int partitionId) {
        this.message.setPartitionId(partitionId);
        return this;
    }

    public MessageBuilder setRandomPartition() {
        this.message.setPartitionId(ConstUtils.PARTITION_RAND);
        return this;
    }

    public MessageBuilder setHashId(long hashId) {
        this.message.setPartitionId(ConstUtils.PARTITION_HASH);
        this.message.setHashId(hashId);
        return this;
    }

    public MessageBuilder setBody(byte[] body) {
        this.message.setBody(body);
        return this;
    }

    public MessageBuilder setBody(String body) {
        this.message.setBody(body.getBytes());
        return this;
    }

    public MessageBuilder setBody(String body, String charsetName) throws UnsupportedEncodingException {
        if (body != null && charsetName != null) {
            this.message.setBody(body.getBytes(charsetName));
        }

        return this;
    }

    public MessageBuilder setKey(String key) {
        if (key != null)
            this.message.setKey(key);
        return this;
    }

    public MessageBuilder setTags(String tag) {
        if (tag != null)
            this.message.setTags(tag);
        return this;
    }

    public MessageBuilder setTraceId(String traceId) {
        this.message.properties.put(producerProxyConstants.TRACE_ID, traceId);
        return this;
    }

    public MessageBuilder setSpanId(String spanId) {
        this.message.properties.put(producerProxyConstants.SPAN_ID, spanId);
        return this;
    }


    public MessageBuilder setPressureTraffic(boolean isOpen) {
        if (isOpen) {
            this.message.properties.put(producerProxyConstants.PRESSURE_TRAFFIC_KEY, producerProxyConstants.PRESSURE_TRAFFIC_ENABLE);
        } else {
            this.message.properties.put(producerProxyConstants.PRESSURE_TRAFFIC_KEY, producerProxyConstants.PRESSURE_TRAFFIC_DISABLE);
        }
        return this;
    }

    public MessageBuilder addProperty(String key, String value) {
        if (key != null && !key.equals(producerProxyConstants.PRESSURE_TRAFFIC_KEY)) {
            this.message.properties.put(key, value);
        }
        return this;
    }

    public Result send() {
        if (this.message.key == null) {
            this.setKey(RandomKeyUtils.randomKey(CarreraConfig.RANDOM_KEY_SIZE));
        }
        return this.producer.sendMessage(this.message);
    }
}